tools/speedy_sync/src/main.rs

```diff
@@ -0,0 +1,272 @@
+        .expect("Failed writing a header");
+
+    store_update
+        .insert_ser(DBCol::BlockInfo, hash.as_ref(), &block_checkpoint.info)
+        .expect("Failed writing a block info");
+
+    store_update
+        .set_ser(DBCol::BlockMerkleTree, hash.as_ref(), &block_checkpoint.merkle_tree)
+        .expect("Failed writing merkle tree");
+    store_update
+        .set_ser(
+            DBCol::BlockHeight,
+            &index_to_bytes(block_checkpoint.header.height()),
+            block_checkpoint.header.hash(),
+        )
+        .unwrap();
+}
+
+fn write_epoch_checkpoint(store_update: &mut StoreUpdate, epoch_checkpoint: &EpochCheckpoint) {
+    store_update
+        .set_ser(DBCol::EpochInfo, epoch_checkpoint.id.as_ref(), &epoch_checkpoint.info)
+        .expect("Failed to write epoch info");
+}
+
+fn create_snapshot(create_cmd: CreateCmd) {
+    let path = Path::new(&create_cmd.home);
+    let store = StoreOpener::with_default_config().read_only(true).home(path).open();
+
+    // Get epoch information:
+    let mut epochs = store
+        .iter(DBCol::EpochInfo)
+        .filter_map(|(key, value)| {
+            if key.as_ref() == AGGREGATOR_KEY {
+                None
+            } else {
+                let info = EpochInfo::try_from_slice(value.as_ref()).unwrap();
+                let id = EpochId::try_from_slice(key.as_ref()).unwrap();
+                Some(EpochCheckpoint { id, info })
+            }
+        })
+        .collect::<Vec<EpochCheckpoint>>();
+
+    assert!(epochs.len() > 4, "Number of epochs must be greater than 4.");
+
+    epochs.sort_by(|a, b| a.info.epoch_height().partial_cmp(&b.info.epoch_height()).unwrap());
+    // Take last two epochs
+    let next_epoch = epochs.pop().unwrap();
+    let current_epoch = epochs.pop().unwrap();
+    let prev_epoch = epochs.pop().unwrap();
+
+    // We need information about 4 blocks to start the chain:
+    //
+    // 'block' - we'll always pick the last block of a given epoch.
+    // 'prev_block' - its predecessor
+    // 'final_block' - the block with finality (usually 2 blocks behind)
+    // 'first_block' - the first block of this epoch (usualy epoch_length behind).
+
+    let block_hash = next_epoch.id.0;
+    let block = read_block_checkpoint(&store, &block_hash);
+    let block_header = block.header.clone();
+    let prev_block = read_block_checkpoint(&store, block_header.prev_hash());
+    let final_block = read_block_checkpoint(&store, block_header.last_final_block());
+    let first_block = read_block_checkpoint(&store, block.info.epoch_first_block());
+
+    let checkpoint = SpeedyCheckpoint {
+        prev_epoch,
+        current_epoch,
+        next_epoch,
+        block,
+        prev_block,
+        final_block,
+        first_block,
+    };
+
+    let serialized = serde_json::to_string(&checkpoint).unwrap();
+
+    fs::write(Path::new(&create_cmd.destination_dir).join("snapshot.json"), serialized)
```

Comment:
Do I understand this right that the JSON is only for debugging / convenience, to have a human-readable copy of the checkpoint? Do you think it will be useful in the future, or was it mostly to develop this feature?

Reply:
originally I wanted to use only JSON -- but I couldn't make it work with u128...
so I've added borsh.

Yes, JSON is mostly useful to have something human readable (as changing anything there will cause the hashes to no longer match).

Reply:
Usually we deserilize `u128` as a string, such as https://github.com/near/nearcore/blob/0042d057af65226d66e08268da43f0b8ac77531b/core/primitives-core/src/serialize.rs#L145-L163.

But I am okay with using Borsh here.